src/D2L.CodeStyle.Analyzers/Async/Generator/SourceFilePathHintNameGenerator.cs

```diff
@@ -0,0 +1,110 @@
+﻿#nullable enable
+
+using System.Collections.Immutable;
+using System.Text.RegularExpressions;
+
+namespace D2L.CodeStyle.Analyzers.Async.Generator {
+	/// <summary>
+	/// Given a stream of input source paths generate a nice looking set of
+	/// distinct hint names for the source generator.
+	/// </summary>
+	internal sealed class SourceFilePathHintNameGenerator {
+		private readonly ImmutableArray<string> m_sourcePaths;
+		public readonly int m_prefixLength;
+
+		private SourceFilePathHintNameGenerator(
+			ImmutableArray<string> sourcePaths,
+			int prefixLength
+		) {
+			m_sourcePaths = sourcePaths;
+			m_prefixLength = prefixLength;
+		}
+
+		public static SourceFilePathHintNameGenerator Create( ImmutableArray<string> sourcePaths ) {
+			// Calculate the longest prefix of all input strings
+
+			if( sourcePaths.Length == 0 ) {
+				return new SourceFilePathHintNameGenerator( sourcePaths, 0 );
+			}
+
+			// Seed it with the first source files directory
+			string commonPrefix = RemoveFileName( sourcePaths[0] );
+
+			foreach( var sourcePath in sourcePaths.Select( RemoveFileName ) ) {
+				int idx = 0;
+				int maxPrefixLength = Math.Min( sourcePath.Length, commonPrefix.Length );
+
+				while( idx < maxPrefixLength ) {
+					if( sourcePath[idx] != commonPrefix[idx] ) {
+						break;
+					}
+
+					idx++;
+				}
+
+				if( idx != commonPrefix.Length ) {
+					commonPrefix = sourcePath.Substring( 0, idx );
+				}
+			}
+
+			return new SourceFilePathHintNameGenerator(
+				sourcePaths,
+				commonPrefix.Length
+			);
+		}
+
+		private static string RemoveFileName( string sourcePath ) {
+			var dir = Path.GetDirectoryName( sourcePath );
+
+			if( sourcePath.Length > dir.Length ) {
+				dir += sourcePath[dir.Length];
+			}
+
+			return dir;
+		}
+
+		private static readonly Regex PathSeparators = new(
+			@"[/\\]",
+			RegexOptions.Compiled
+		);
+
+		private static readonly Regex DisallowedPathChars = new(
+			@"[^a-zA-Z0-9-._]*",
+			RegexOptions.Compiled
+		);
+
+		public IEnumerable<string> GetHintNames() {
+			// Hint names need to be unique, but the sanitization we could do
+			// may introduce ambiguities.
+			var used = new HashSet<string>();

```

Comment:
Alternatives we talked about:

* When there is an ambiguity suffix with the sha256 of the path or something: a pro here is that it would be more stable as people add different paths (the user of this code will pass in sorted input, but if you add/remove files then it could cause unrelated files to have different hintnames) which would be interesting if we were checking the output into git (so that it shows up in PRs with diffs)
* We talked about an unambiguous encoding, like how percent encoding does ☃ -> `%E2%98%83` and `%` -> `%25` or other related schemes... but then I'd need to not do the `.Replace( x, "" )` below that deletes bad chars.

Honestly I don't really care, I'll do whatever. I don't think this is a very likely scenario (maybe if you had `D2L.Foo.Bar\x.cs` and `D2L.FooBar\x.cs`? but like realy?) but I didn't want to crash (really annoying in VS) or bother plumbing diagnostics out of here.